### PR TITLE
[sentinel] Adds group by repository functionality

### PR DESCRIPTION
**Pull request by Numbers88s**
Adds transport and store layer for retrieving the grouping of vulnerability matches by repository.

CleanShot 2023-03-24 at 10 11 48

## Test plan
Tested manually. Adding tests shortly

### CODE REVIEW

File: enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go [hunk 1]
```diff
@@ -358,3 +390,53 @@ func (r *vulnerabilityMatchResolver) PreciseIndex(ctx context.Context) (resolver
...
+}
+
+func (v vulnerabilityMatchGroupByRepositoryResolver) MatchCount() int32 {
+	return v.v.MatchCount
+}
+
+//
+//
+
+type vulnerabilityMatchGroupByRepositoryConnectionResolver struct {
```
Comment (efritz) [inline enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go:416]: We have new shared graphql utilities that abstracts connections that do the same offset/limit pagination. (we could also apply these new things in a follow-up PR)

File: cmd/frontend/graphqlbackend/codeintel.sentinel.graphql [hunk 2]
```diff
@@ -254,3 +274,42 @@ type VulnerabilityMatch implements Node {
...
+    """
+    Information on how to fetch the next page.
+    """
+    pageInfo: PageInfo!
+}
+
+"""
+Vulnerability matches count group by repository.
+"""
+type VulnerabilityMatchGroupByRepository {
```
Comment (efritz) [inline cmd/frontend/graphqlbackend/codeintel.sentinel.graphql:301]: Looks like you're using `GroupBy` for the graphql layer but `CountBy` for the database layer stuff. I think using the latter consistent would make more sense - not sure why I was "grouping" in the graphql layer at first.

File: enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go [hunk 3]
```diff
@@ -149,6 +151,36 @@ func (r *rootResolver) VulnerabilityMatches(ctx context.Context, args resolverst
 	return resolverstubs.NewTotalCountConnectionResolver(resolvers, offset, int32(totalCount)), nil
 }
 
+func (r *rootResolver) VulnerabilityMatchesGroupByRepository(ctx context.Context, args resolverstubs.GetVulnerabilityMatchGroupByRepositoryArgs) (_ resolverstubs.VulnerabilityMatchGroupByRepositoryConnectionResolver, err error) {
+	ctx, _, endObservation := r.operations.vulnerabilityMatchesGroupByRepository.WithErrors(ctx, &err, observation.Args{LogFields: []log.Field{}})
+	endObservation.OnCancel(ctx, 1, observation.Args{})
+
+	limit, offset, err := args.ParseLimitOffset(50)
```
Comment (Numbers88s) [inline enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go:158]: I did use the new abstracted limit offset helpers. Is there something else I am missing?

File: enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go (hunk 1 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go:416]: https://github.com/sourcegraph/sourcegraph/pull/49923/files#r1147732822

File: cmd/frontend/graphqlbackend/codeintel.sentinel.graphql (hunk 2 above)
Comment (Numbers88s) [reply cmd/frontend/graphqlbackend/codeintel.sentinel.graphql:301]: Oh yeah, I just noticed it. Thanks! Changing it here shortly.

File: enterprise/internal/codeintel/sentinel/internal/store/matches.go [hunk 4]
```diff
@@ -127,6 +127,54 @@ var flattenMatches = func(ms []shared.VulnerabilityMatch) []shared.Vulnerability
...
 }
 
+// GetVulnerabilityMatchesCountByRepository returns the number vulnerabilities matches count by repository
+func (s *store) GetVulnerabilityMatchesCountByRepository(ctx context.Context, args shared.GetVulnerabilityMatchesGroupByRepositoryArgs) (_ []shared.VulnerabilityMatchesByRepository, _ int, err error) {
+	ctx, _, endObservation := s.operations.getVulnerabilityMatchesCountByRepository.With(ctx, &err, observation.Args{})
+	defer endObservation(1, observation.Args{})
+
+	var conds []*sqlf.Query
+	if args.RepositoryName != "" {
+		conds = append(conds, sqlf.Sprintf("r.name ILIKE %s", "%"+args.RepositoryName+"%"))
```
Comment (Numbers88s) [inline enterprise/internal/codeintel/sentinel/internal/store/matches.go:137]: It's using ILIKE

File: cmd/frontend/graphqlbackend/codeintel.sentinel.graphql [hunk 5]
```diff
@@ -43,6 +43,26 @@ extend type Query {
...
+
+        """
+        If supplied, indicates which results to skip over during pagination.
+        """
+        after: String
+
+        """
+        The name of the repository to group by.
+        """
+        repositoryName: String
```
[the remark this answers is not in the document]
Comment (Numbers88s) [reply cmd/frontend/graphqlbackend/codeintel.sentinel.graphql:64]: it doesn't need to match exactly.

See [here](https://github.com/sourcegraph/sourcegraph/pull/49923/files#r1147795203)

[the remark this answers is not in the document]
Comment (efritz) [reply cmd/frontend/graphqlbackend/codeintel.sentinel.graphql:64]: Can you make that clear in this docstring?

File: enterprise/internal/codeintel/sentinel/internal/store/matches.go [hunk 6]
```diff
@@ -127,6 +127,54 @@ var flattenMatches = func(ms []shared.VulnerabilityMatch) []shared.Vulnerability
...
+
+const getVulnerabilityMatchesGroupedByRepos = `
+select
+	r.id,
+	r.name,
+	count(*) as count,
+	COUNT(*) OVER() AS total_count
+from vulnerability_matches vm
+join lsif_uploads lu on vm.upload_id = lu.id
+join repo r on lu.repository_id = r.id
```
Comment (efritz) [inline enterprise/internal/codeintel/sentinel/internal/store/matches.go:171]:
```suggestion
join lsif_uploads lu on lu.id = vm.upload_id
join repo r on r.id = lu.repository_id
```

better to have `JOIN T t ON t.X = y` as a pattern (left side of comparison is always the join target, not the context)

File: enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go (hunk 3 above)
Comment (efritz) [reply enterprise/internal/codeintel/sentinel/transport/graphql/root_resolver.go:158]: Some other utils you might find handy:

- [`NewConnectionResolver`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@09c08c6f429c38f844ea973c9e8c8f8b80109c6d/-/blob/internal/codeintel/resolvers/utils.go?L19)
- [`NewTotalCountConnectionResolver`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@09c08c6f429c38f844ea973c9e8c8f8b80109c6d/-/blob/internal/codeintel/resolvers/utils.go?L116)
- [`MarshalID`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@09c08c6f429c38f844ea973c9e8c8f8b80109c6d/-/blob/internal/codeintel/resolvers/utils.go?L231)

File: cmd/frontend/graphqlbackend/codeintel.sentinel.graphql (hunk 5 above)
[the remark this answers is not in the document]
Comment (Numbers88s) [reply cmd/frontend/graphqlbackend/codeintel.sentinel.graphql:64]: Yup.

**Pull request merged**